ambari-server/src/main/java/org/apache/ambari/server/stack/upgrade/ColocatedGrouping.java

```diff
@@ -93,75 +98,107 @@ private MultiHomedBuilder(Grouping grouping, Batch batch, boolean serviceCheck)
     public void add(UpgradeContext context, HostsType hostsType, String service,
         boolean clientOnly, ProcessingComponent pc, Map<String, String> params) {
 
+      // !!! the percent is the number of co-located items that should be run
+      // before pausing. This value has no bearing on how many should be in a single stage.
       int count = Double.valueOf(Math.ceil(
           (double) m_batch.percent / 100 * hostsType.getHosts().size())).intValue();
 
-      int i = 0;
-      for (String host : hostsType.getHosts()) {
-        // This class required inserting a single host into the collection
-        HostsType singleHostsType = HostsType.single(host);
+      LinkedHashSet<String> first = new LinkedHashSet<>();
+      LinkedHashSet<String> remaining = new LinkedHashSet<>();
+      hostsType.getHosts().stream().forEach(hostName -> {
+        if (first.size() < count) {
+          first.add(hostName);
+        } else {
+          remaining.add(hostName);
+        }
+      });
+
+      // !!! resolve tasks only once
+      List<Task> preTasks = resolveTasks(context, true, pc);
+      Task task = resolveTask(context, pc);
+      List<Task> postTasks = resolveTasks(context, false, pc);
+      AtomicBoolean processInitial = new AtomicBoolean(true);
+
+      int parallelCount = -1;
+      OrchestrationOptions options = context.getOrchestrationOptions();
+      if (null != options) {
+        parallelCount = context.getOrchestrationOptions().getConcurrencyCount(
+            context.getCluster().buildClusterInformation(), service, pc.name);
+      }
 
-        Map<String, List<TaskProxy>> targetMap = ((i++) < count) ? initialBatch : finalBatches;
-        List<TaskProxy> targetList = targetMap.get(host);
+      if (parallelCount < 1) {
+        parallelCount = getParallelHostCount(context, 1);
+      }
 
-        if (null == targetList) {
-          targetList = new ArrayList<>();
-          targetMap.put(host, targetList);
-        }
+      // !!! stupid effective-final check
+      final int hostCount = parallelCount;
```

Comment:
You wouldn't need to do this if you didn't use all those fancy streams() :)

Reply:
but it's _modern_ lol